src/openms/source/ANALYSIS/OPENSWATH/MRMScoring.cpp

```diff
@@ -705,122 +673,131 @@ namespace OpenSwath
-          mi_contrast_matrix_.setValue(i, j, Scoring::rankedMutualInformation(intensityi, intensityj));
+          mi_contrast_matrix_.setValue(i, j, Scoring::rankedMutualInformation(rank_vec1[i], rank_vec2[j]));
         }
       }
     }
 
     void MRMScoring::initializeMIPrecursorMatrix(OpenSwath::IMRMFeature* mrmfeature, const std::vector<String>& precursor_ids)
     {
-      std::vector<double> intensityi, intensityj;
+      std::vector<std::vector<double>> intensity;
+      std::vector<std::vector<unsigned int>> rank_vec;
+      fillIntensityFromPrecursorFeature(mrmfeature, precursor_ids, intensity);
+      Scoring::computeRankVector(intensity, rank_vec);
+
       mi_precursor_matrix_.resize(precursor_ids.size(),precursor_ids.size());
       for (std::size_t i = 0; i < precursor_ids.size(); i++)
       {
-        FeatureType fi = mrmfeature->getPrecursorFeature(precursor_ids[i]);
-        intensityi.clear();
-        fi->getIntensity(intensityi);
         for (std::size_t j = i; j < precursor_ids.size(); j++)
         {
-          FeatureType fj = mrmfeature->getPrecursorFeature(precursor_ids[j]);
-          intensityj.clear();
-          fj->getIntensity(intensityj);
           // compute ranked mutual information
-          mi_precursor_matrix_.setValue(i, j, Scoring::rankedMutualInformation(intensityi, intensityj));
+          mi_precursor_matrix_.setValue(i, j, Scoring::rankedMutualInformation(rank_vec[i], rank_vec[j]));
         }
       }
     }
 
     void MRMScoring::initializeMIPrecursorContrastMatrix(OpenSwath::IMRMFeature* mrmfeature, const std::vector<String>& precursor_ids, const std::vector<String>& native_ids)
     {
-      std::vector<double> intensityi, intensityj;
+      std::vector<std::vector<double>> intensityi, intensityj;
+      std::vector<std::vector<unsigned int>> rank_vec1, rank_vec2;
+      fillIntensityFromPrecursorFeature(mrmfeature, precursor_ids, intensityi);
+      fillIntensityFromFeature(mrmfeature, native_ids, intensityj);
+      Scoring::computeRankVector(intensityi, rank_vec1);
+      Scoring::computeRankVector(intensityj, rank_vec2);
+
       mi_precursor_contrast_matrix_.resize(precursor_ids.size(), native_ids.size());
       for (std::size_t i = 0; i < precursor_ids.size(); i++)
       {
-        FeatureType fi = mrmfeature->getPrecursorFeature(precursor_ids[i]);
-        //mi_precursor_contrast_matrix_[i].resize(native_ids.size());
-        intensityi.clear();
-        fi->getIntensity(intensityi);
         for (std::size_t j = 0; j < native_ids.size(); j++)
         {
-          FeatureType fj = mrmfeature->getFeature(native_ids[j]);
-          intensityj.clear();
-          fj->getIntensity(intensityj);
           // compute ranked mutual information
-          mi_precursor_contrast_matrix_.setValue(i, j, Scoring::rankedMutualInformation(intensityi, intensityj));
+          mi_precursor_contrast_matrix_.setValue(i, j, Scoring::rankedMutualInformation(rank_vec1[i], rank_vec2[j]));
```

Comment:
this computes a full matrix. can you safe time by computing only the upper triangle matrix * 2 + diagonal?

Reply:
This matrix is not symmetric so the full matrix has to be computed